lightning/src/ln/wire.rs

```diff
@@ -0,0 +1,411 @@
+//! Wire encoding/decoding for Lightning messages according to [BOLT #1].
+//!
+//! Messages known by this module can be read from the wire using [`read`].
+//! The [`Message`] enum returned by [`read`] wraps the decoded message or the message type (if
+//! unknown) to use with pattern matching.
+//!
+//! Messages implementing the [`Encode`] trait define a message type and can be sent over the wire
+//! using [`write`].
+//!
+//! [BOLT #1]: https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md
+//! [`read`]: fn.read.html
+//! [`write`]: fn.write.html
+//! [`Encode`]: trait.Encode.html
+//! [`Message`]: enum.Message.html
+
+use ln::msgs;
+use util::ser::{Readable, Writeable, Writer};
+
+/// A Lightning message returned by [`read`] when decoding bytes received over the wire. Each
+/// variant contains a message from [`ln::msgs`] or otherwise the message type if unknown.
+///
+/// [`read`]: fn.read.html
+/// [`ln::msgs`]: ../msgs/index.html
+#[allow(missing_docs)]
+pub enum Message {
+	Init(msgs::Init),
+	Error(msgs::ErrorMessage),
+	Ping(msgs::Ping),
+	Pong(msgs::Pong),
+	OpenChannel(msgs::OpenChannel),
+	AcceptChannel(msgs::AcceptChannel),
+	FundingCreated(msgs::FundingCreated),
+	FundingSigned(msgs::FundingSigned),
+	FundingLocked(msgs::FundingLocked),
+	Shutdown(msgs::Shutdown),
+	ClosingSigned(msgs::ClosingSigned),
+	UpdateAddHTLC(msgs::UpdateAddHTLC),
+	UpdateFulfillHTLC(msgs::UpdateFulfillHTLC),
+	UpdateFailHTLC(msgs::UpdateFailHTLC),
+	UpdateFailMalformedHTLC(msgs::UpdateFailMalformedHTLC),
+	CommitmentSigned(msgs::CommitmentSigned),
+	RevokeAndACK(msgs::RevokeAndACK),
+	UpdateFee(msgs::UpdateFee),
+	ChannelReestablish(msgs::ChannelReestablish),
+	AnnouncementSignatures(msgs::AnnouncementSignatures),
+	ChannelAnnouncement(msgs::ChannelAnnouncement),
+	NodeAnnouncement(msgs::NodeAnnouncement),
+	ChannelUpdate(msgs::ChannelUpdate),
+	/// A message that could not be decoded because its type is unknown.
+	Unknown(MessageType),
+}
+
+/// A number identifying a message to determine how it is encoded on the wire.
+#[derive(Clone, Copy)]
+pub struct MessageType {
```

Comment:
Could probably be MessageType(u16), no?

Reply:
Good point! Seems it's a simple enough type to warrant a tuple struct. I've updated it in 298a7d8.

Initially, I had given it both a number and an even/odd parity. But I didn't like how I had to expose the fields as public to match on the parity. And if they were public then they could be modified such that the number and parity did not agree. So I switch to using an `is_even` method for parity.